### PR TITLE
Add scripts to generate a single PDF from the documentation

### PR DESCRIPTION
**Pull request by szarnyasg**
This PR contains scripts that produce a single PDF from the documentation using Pandoc and XeLaTeX. The 

The `single-file-document/README.md` contains instructions on how to use the scripts from the shell natively or via Docker.

### CODE REVIEW

**Review by Mause — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 1]
```diff
@@ -82,7 +82,7 @@ build/
 *.lo
 *.la
 *~
-*.pdf
+/*.pdf
```
[the remark this answers is not in the document]
Comment (szarnyasg) [reply .gitignore:85]: Yes - the need other PDFs in the repo such as the emojis in `single-file-document/emojis/`.

File: single-file-document/requirements.txt [hunk 2]
```diff
@@ -0,0 +1 @@
+PyYAML
```
[the remark this answers is not in the document]
Comment (szarnyasg) [reply single-file-document/requirements.txt:1]: Exactly, we need both

File: single-file-document/concatenate_to_single_file.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        
+        of.write(f" {{#{path_label}}}\n")
+
+        # move headers h2-h4 down by 3 levels (to h5-h7)
+        extra_header_levels = 3*"#"
+        doc_body = re.sub(r"^##", f"##{extra_header_levels}", doc_body, flags=re.MULTILINE)
+
+        # replace blog post paths to the full URL
+        doc_body = re.sub(
+                r"\(/(20[0-9][0-9]/[0-9][0-9]/[0-9][0-9])/",
```
[the remark this answers is not in the document]
Comment (szarnyasg) [reply single-file-document/concatenate_to_single_file.py:59]: This looks for links that immediately start with a slash, e.g. `[ link text ]<<(/2023-08-15/>>blog-post.html)` where the part between `<< ... >>` is the match.

File: single-file-document/concatenate_to_single_file.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+
+        # drop "Pages in this Section" and "More" sections
+        doc_body = doc_body.replace("### Pages in this Section", "")
+        doc_body = doc_body.replace("### More", "")
+
+        # drop lines containing "---", pandoc interprets these as h2 headers
+        doc_body = re.sub(r"^---$", "", doc_body, flags=re.MULTILINE)
+
+        # add path labels to headers at the beginning of the file
+        path_label = full_path \
```
[the remark this answers is not in the document]
Comment (szarnyasg) [reply single-file-document/concatenate_to_single_file.py:45]: Yes, e.g.

```md
# Hello {#my%world}
```
Gets translated to:
```html
<h1 id="hello-myworld">Hello {#my%world}</h1>
```
with the `%` removed. That said, we could do our own sanitization to be sure that we only use characters in`[-:a-zA-Z0-9]` for the labels.

File: single-file-document/concatenate_to_single_file.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+        doc_title = doc["title"]
+        doc_body = doc.content
+
+        of.write(f"""{"#" * header_level} {doc_title}""")
+
+        # drop "Pages in this Section" and "More" sections
+        doc_body = doc_body.replace("### Pages in this Section", "")
+        doc_body = doc_body.replace("### More", "")
+
+        # drop lines containing "---", pandoc interprets these as h2 headers
```
[the remark this answers is not in the document]
Comment (szarnyasg) [reply single-file-document/concatenate_to_single_file.py:41]: The documentation uses them here:
https://github.com/duckdb/duckdb-web/blob/master/docs/api/c/api.md#duckdb_open_ext
Like so:
```md
### `duckdb_open`
---
Creates a new database or opens an existing database file stored at the the given path.
If no path is given a new in-memory database is created instead.
The instantiated database should be closed with 'duckdb_close'

#### Syntax
---
<div class="language-c highlighter-rouge"><div class="highlight"><pre class="highlight"><code><span class="kt">duckdb_state</span> <span class="k">duckdb_open</span>(<span class="k">
</span>  <span class="kt">const</span> <span class="kt">char</span> *<span class="k">path</span>,<span class="k">
</span>  <span class="kt">duckdb_database</span> *<span class="k">out_database
</span>);
</code></pre></div></div>
#### Parameters
---
* `path`
```

If we indeed interpret them as horizontal breaks, we'll get this:
Screenshot 2023-08-15 at 16 10 49

This does not seem right to me. (Let's ignore for the time being that the `Syntax` section, which is pure HTML is not yet rendered.)

File: single-file-document/concatenate_to_single_file.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+import textwrap
+from datetime import datetime, timezone
+import argparse
+import pathlib
+import frontmatter
+
+
+def path_to_label(docs_root_absolute_path, path):
+    resolved_path = str(pathlib.Path(str(docs_root_absolute_path) + '/' + path).resolve().relative_to(docs_root_absolute_path))
+    resolved_path = re.sub(r"\.md$", "", resolved_path)
```
[the remark this answers is not in the document]
Comment (szarnyasg) [reply single-file-document/concatenate_to_single_file.py]: I simplified it to use `os.path.relpath` and added a lot of comments to explain what it actually does.

File: single-file-document/concatenate_to_single_file.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+    resolved_path = resolved_path.replace("/", ":")
+    return resolved_path
+
+
+def concat(of, header_level, docs_root_absolute_path, docs_root, doc_path):
```
[the remark this answers is not in the document]
Comment (szarnyasg) [reply single-file-document/concatenate_to_single_file.py]: Yes, 1b3677044b193729e0a99601978d6cd5fcbc45e3 implements this.

**Comment by carlopi** [comment]
One question I have, not strictly related to this PR, is about naming of the single file documentation.

I would consider having the name encode versioning somehow, even just having `duckdb-latest.pd`f vs `duckdb-0.8.1.pdf` vs `duckdb-0.9.0.pdf`, so that when you download them you can still tell them apart from the name.

And I would consider (later!) whether they should be available to be downloaded via SQL (this requires both some PRAGMA and deciding where to store them, somewhere inside `~/.duckdb/` like `~/.duckdb/docs/duckdb-0.8.1.pdf` or equivalent).

**Pull request merged**